webrtc-c/canary/src/Peer.cpp

```diff
@@ -0,0 +1,191 @@
+
+namespace Canary {
+
+Peer::Peer(const Canary::PConfig pConfig, const Callbacks& callbacks)
+    : pConfig(pConfig), callbacks(callbacks), pAwsCredentialProvider(nullptr), terminated(FALSE)
+{
+}
+
+Peer::~Peer()
+{
+    freeStaticCredentialProvider(&pAwsCredentialProvider);
+}
+
+STATUS Peer::init()
+{
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK_STATUS(createStaticCredentialProvider((PCHAR) pConfig->pAccessKey, 0, (PCHAR) pConfig->pSecretKey, 0, (PCHAR) pConfig->pSessionToken, 0,
+                                              MAX_UINT64, &pAwsCredentialProvider));
+
+CleanUp:
+
+    return retStatus;
+}
+
+VOID Peer::shutdown()
+{
+    this->terminated = TRUE;
+    CHK_LOG_ERR(freeSignalingClient(&this->pSignalingClientHandle));
+
+    for (auto& connection : this->connections) {
+        connection->shutdown();
+    }
+
+    CHK_LOG_ERR(freeStaticCredentialProvider(&this->pAwsCredentialProvider));
+}
+
+STATUS Peer::connect()
+{
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK_STATUS(connectSignaling());
+
+    if (!this->pConfig->isMaster) {
+        std::shared_ptr<Connection> pConnection;
+        CHK_STATUS(this->createNewConnection(DEFAULT_VIEWER_PEER_ID, &pConnection));
+        CHK_STATUS(pConnection->connect());
+    }
+
+CleanUp:
+
+    return retStatus;
+}
+
+STATUS Peer::connectSignaling()
+{
+    STATUS retStatus = STATUS_SUCCESS;
+
+    SignalingClientInfo clientInfo;
+    ChannelInfo channelInfo;
+    SignalingClientCallbacks clientCallbacks;
+
+    MEMSET(&clientInfo, 0, SIZEOF(clientInfo));
+    MEMSET(&channelInfo, 0, SIZEOF(channelInfo));
+    MEMSET(&clientCallbacks, 0, SIZEOF(clientCallbacks));
+
+    clientInfo.version = SIGNALING_CLIENT_INFO_CURRENT_VERSION;
+    clientInfo.loggingLevel = pConfig->logLevel;
+    STRCPY(clientInfo.clientId, pConfig->pClientId);
+
+    channelInfo.version = CHANNEL_INFO_CURRENT_VERSION;
+    channelInfo.pChannelName = (PCHAR) pConfig->pChannelName;
+    channelInfo.pKmsKeyId = NULL;
+    channelInfo.tagCount = 0;
+    channelInfo.pTags = NULL;
+    channelInfo.channelType = SIGNALING_CHANNEL_TYPE_SINGLE_MASTER;
+    channelInfo.channelRoleType = pConfig->isMaster ? SIGNALING_CHANNEL_ROLE_TYPE_MASTER : SIGNALING_CHANNEL_ROLE_TYPE_VIEWER;
+    channelInfo.cachingPolicy = SIGNALING_API_CALL_CACHE_TYPE_FILE;
+    channelInfo.cachingPeriod = SIGNALING_API_CALL_CACHE_TTL_SENTINEL_VALUE;
+    channelInfo.asyncIceServerConfig = TRUE;
+    channelInfo.retry = TRUE;
+    channelInfo.reconnect = TRUE;
+    channelInfo.pCertPath = (PCHAR) pConfig->pCertPath;
+    channelInfo.messageTtl = 0; // Default is 60 seconds
+
+    clientCallbacks.customData = (UINT64) this;
+    clientCallbacks.errorReportFn = [](UINT64 customData, STATUS status, PCHAR msg, UINT32 msgLen) -> STATUS {
+        PPeer pPeer = (PPeer) customData;
+        DLOGW("Signaling client generated an error 0x%08x - '%.*s'", status, msgLen, msg);
+
+        UNUSED_PARAM(pPeer);
+        // TODO: handle recreate signaling
```

Comment:
This is a big one

Reply:
Instead of recreating the signaling, I went with the "let it crash" philosophy, and capture the error code as a part of metrics.